dgl_ptm/tests/test_agent.py

```diff
@@ -0,0 +1,212 @@
+        # assert NotImplementedError was raised
+        with pytest.raises(NotImplementedError):
+            agent_perception_update(
+                model.model_graph,
+                model.model_data,
+                1,
+                method='other'
+                )
+
+
+class TestCapitalUpdate:
+
+    def test_capital_update_default(self, model):
+
+        k = model.model_graph.ndata['wealth']
+        c = model.model_graph.ndata['wealth_consumption']
+        i_a = model.model_graph.ndata['i_a']
+        m = model.model_graph.ndata['m']
+        global_θ =model.model_data['modelTheta'][0]
+        𝛿 = model.steering_parameters['depreciation']
+        # generate random income with size 100
+        income = torch.rand(100)
+        expected_wealth = (global_θ + m * (1-global_θ)) * (income - c - i_a + (1-𝛿) * k)
+
+        # because income_generation is not called yet
+        model.model_graph.ndata['income'] = income
+
+        capital_update(
+            model.model_graph,
+            model.steering_parameters,
+            model.model_data,
+            1,
+            method="default"
+            )
+
+        # assert _agent_capital_update was called
+        assert (model.model_graph.ndata['wealth'] == expected_wealth).all()
+
+    def test_capital_update_other(self, model):
+        # assert NotImplementedError was raised
+        with pytest.raises(NotImplementedError):
+            capital_update(
+                model.model_graph,
+                model.model_data,
+                1,
+                method='other'
+                )
+
+
+class TestIncomeGeneration:
+
+    def test_income_generation_default(self, model):
+        gamma = model.steering_parameters['tech_gamma']
+        cost = model.steering_parameters['tech_cost']
+        alpha = model.model_graph.ndata['alpha']
+        wealth = model.model_graph.ndata['wealth']
+        expected_income,expected_tech_index = torch.max((alpha[:,None]*wealth[:,None]**gamma - cost), axis=1)
+
+        income_generation(
+            model.model_graph,
+            model.device,
+            model.steering_parameters,
+            method="default"
+            )
+
+        # assert _agent_income_generator was called
+        assert (model.model_graph.ndata['income'] == expected_income).all()
+        assert (model.model_graph.ndata['tech_index'] == expected_tech_index).all()
+
+    def test_income_generation_other(self, model):
+        # assert NotImplementedError was raised
+        with pytest.raises(NotImplementedError):
+            income_generation(
+                model.model_graph,
+                model.device,
+                model.steering_parameters,
+                method='other'
+                )
+
+
+class TestWealthConsumption:
+
+    def test_wealth_consumption_default(self, model):
+        wealth = model.model_graph.ndata['wealth']
+        expected_wealth_consumption = 0.64036047*torch.log(wealth)
```

Comment:
Maybe we can document the origin of the numerical value here (as well as in the actual function)

Reply:
Good point. I couldn't find a reference, but submitted issue #83 and wrote it as a todo in the actual function.